api/paidAction/index.js

```diff
@@ -30,75 +34,166 @@ export const paidActions = {
   TERRITORY_BILLING,
   TERRITORY_UNARCHIVE,
   DONATE,
+  TRANSFER,
   BUY_CREDITS
 }
 
+async function checkUser (userId, { models, tx, me }, minSats, minCreds) {
+  // count pending invoices and bail if we're over the limit
+  const pendingInvoices = await ((tx ?? models).invoice.count({
+    where: {
+      userId,
+      actionState: {
+        // not in a terminal state. Note: null isn't counted by prisma
+        notIn: PAID_ACTION_TERMINAL_STATES
+      }
+    }
+  }))
+
+  console.log('pending paid actions', pendingInvoices)
+  if (pendingInvoices >= MAX_PENDING_PAID_ACTIONS_PER_USER) {
+    throw new Error('you have too many pending paid actions, cancel some or wait for them to expire')
+  }
+
+  if (minSats != null && toPositiveBigInt(minSats) > (me?.msats ?? 0n)) {
+    throw new Error('you do not have enough sats to perform this action')
+  }
+  if (minCreds != null && toPositiveBigInt(minCreds) > (me?.mcredits ?? 0n)) {
+    throw new Error('you do not have enough credits to perform this action')
+  }
+}
+
 export default async function performPaidAction (actionType, args, context) {
   try {
-    const { me, models, forceFeeCredits } = context
-    const paidAction = paidActions[actionType]
-
+    context = { ...context }
     console.group('performPaidAction', actionType, args)
-
+    const paidAction = paidActions[actionType]
     if (!paidAction) {
-      throw new Error(`Invalid action type ${actionType}`)
+      throw new Error(`ynvalid action type ${actionType}`)
     }
 
-    if (!me && !paidAction.anonable) {
-      throw new Error('You must be logged in to perform this action')
-    }
+    context.me = context.me ? await context.models.user.findUnique({ where: { id: context.me.id } }) : undefined
+    context.actionAttempt = context.actionAttempt ?? 0
+    context.forceInternal = context.forceInternal ?? false
+    context.prioritizeInternal = context.prioritizeInternal ?? false
+
+    context.cost = await paidAction.getCost(args, context) ?? 0n
+    if (context.cost < 0n) throw new Error('cost cannot be negative')
 
-    context.me = me ? await models.user.findUnique({ where: { id: me.id } }) : undefined
-    context.cost = await paidAction.getCost(args, context)
+    context.sybilFeePercent = paidAction.getSybilFeePercent ? await paidAction.getSybilFeePercent(args, context) : undefined
+    if (context.sybilFeePercent !== undefined && context.sybilFeePercent < 0n) throw new Error('sybil fee percent cannot be negative')
 
-    // special case for zero cost actions
-    if (context.cost === 0n) {
+    context.description = context.me?.hideInvoiceDesc ? undefined : await paidAction.describe(args, context)
+    context.descriptionHash = paidAction.describeHash ? await paidAction.describeHash(args, context) : undefined
+    context.supportedPaymentMethods = paidAction.paymentMethods ?? (paidAction.getPaymentMethods ? await paidAction.getPaymentMethods(args, context) : undefined)
```

Comment:
Optionally actions can export a getter: `getPaymentMethods` that is used if `paymentMethod` field is not exported.  Maybe we can change payment priorities for some actions depending on user preference?